## **1) Preprossessing**

### **Transformations for Customer Data table**

In [1]:
-- Check data types
SELECT COLUMN_NAME, DATA_TYPE
FROM INFORMATION_SCHEMA.COLUMNS
WHERE TABLE_NAME = 'WOODCORP_ETC_CUSTOMERDATA';

(5 rows affected)

Total execution time: 00:00:00.134

COLUMN_NAME DATA_TYPE CUSTOMER_NO int LOYALTY_PROGRAM tinyint AGE tinyint CUSTOMER_TYPE tinyint GENDER tinyint

No need to chane anything:  
\- All in tinyint works given that all the columns in tinyint have values between 0 and 4.  
<span style="font-family: -apple-system, BlinkMacSystemFont, sans-serif; color: var(--vscode-foreground);">- Int for the primary key (Cutomer_NO) is fine.</span>

In [2]:
-- Check if Primary Keys Exist
SELECT 
    column_name
FROM 
    information_schema.key_column_usage
WHERE 
    table_name = 'WOODCORP_ETC_CUSTOMERDATA' 
    AND constraint_name LIKE 'PK_%';

(0 rows affected)

Total execution time: 00:00:00.136

column_name

No primary key. Let's set it!

In [3]:
-- Drop existing primary key constraint if any
IF EXISTS (SELECT * FROM information_schema.table_constraints
           WHERE table_name = 'WOODCORP_ETC_CUSTOMERDATA'
           AND constraint_type = 'PRIMARY KEY')
BEGIN
    DECLARE @sql NVARCHAR(MAX)
    SET @sql = (SELECT 'ALTER TABLE WOODCORP_ETC_CUSTOMERDATA DROP CONSTRAINT ' + constraint_name
                FROM information_schema.table_constraints
                WHERE table_name = 'WOODCORP_ETC_CUSTOMERDATA'
                AND constraint_type = 'PRIMARY KEY')
    EXEC sp_executesql @sql
END

-- Add primary key constraint on CUSTOMER_NO
ALTER TABLE WOODCORP_ETC_CUSTOMERDATA
ADD CONSTRAINT WOODCORP_ETC_CUSTOMERDATA_PK PRIMARY KEY (CUSTOMER_NO);


: Msg 3725, Level 16, State 0, Line 1
The constraint 'WOODCORP_ETC_CUSTOMERDATA_PK' is being referenced by table 'WOODCORP_ETC_OFFLINE', foreign key constraint 'WOODCORP_ETC_OFFLINE_FK'.

: Msg 3727, Level 16, State 0, Line 1
Could not drop constraint. See previous errors.

Total execution time: 00:00:00.121

### **Transformations for Online Table**

In [60]:
-- Check data types
SELECT COLUMN_NAME, DATA_TYPE
FROM INFORMATION_SCHEMA.COLUMNS
WHERE TABLE_NAME = 'WOODCORP_ETC_ONLINE';

(6 rows affected)

Total execution time: 00:00:00.478

COLUMN_NAME DATA_TYPE ORDER_NO int CUSTOMER_NUMBER_ONLINE int ORDER_VALUE_ONLINE float DISCOUNT_VALUE_ONLINE float NUMBER_ITEMS_ONLINE int NUMBER_DISCOUNTED_ITEMS_ONLINE int

All is good no need to change!

In [64]:
-- Check if Primary Keys Exist
SELECT 
    kcu.COLUMN_NAME
FROM 
    INFORMATION_SCHEMA.TABLE_CONSTRAINTS tc
    JOIN INFORMATION_SCHEMA.KEY_COLUMN_USAGE kcu
        ON tc.CONSTRAINT_NAME = kcu.CONSTRAINT_NAME
WHERE 
    tc.TABLE_NAME = 'WOODCORP_ETC_ONLINE'
    AND tc.CONSTRAINT_TYPE = 'PRIMARY KEY';

(0 rows affected)

Total execution time: 00:00:00.539

COLUMN_NAME

In [65]:
-- Check if foreign Keys Exist
SELECT 
    kcu.COLUMN_NAME
FROM 
    INFORMATION_SCHEMA.TABLE_CONSTRAINTS tc
    JOIN INFORMATION_SCHEMA.KEY_COLUMN_USAGE kcu
        ON tc.CONSTRAINT_NAME = kcu.CONSTRAINT_NAME
WHERE 
    tc.TABLE_NAME = 'WOODCORP_ETC_ONLINE'
    AND tc.CONSTRAINT_TYPE = 'FOREIGN KEY';

(0 rows affected)

Total execution time: 00:00:00.504

COLUMN_NAME

No keys defined! Let's add them

Attempted to set primary keys but encountered issues due to existing primary key constraint. There are 600 unmatched foreign key values in the CUSTOMER\_NUMBER\_ONLINE column that do not match any value in the CUSTOMER\_NO column of the WOODCORP\_ETC\_CUSTOMERDATA table. This indicates there are orders with no corresponding customer information.Let's find and remove these unmatched records as they hinder our ability to perform accurate analysis.

In [69]:
-- Find unmatched foreign key values
SELECT DISTINCT CUSTOMER_NUMBER_ONLINE 
FROM WOODCORP_ETC_ONLINE
WHERE CUSTOMER_NUMBER_ONLINE NOT IN (SELECT CUSTOMER_NO FROM WOODCORP_ETC_CUSTOMERDATA);


(606 rows affected)

Total execution time: 00:00:00.444

CUSTOMER_NUMBER_ONLINE 53844486 53879355 53933388 53955729 53970765 53973495 54002796 54028599 54096585 54096627 54122538 54142452 54174357 54210612 54232230 54240480 54269130 54293781 54302181 54303471 54338472 54357336 54358911 54368097 54368106 54383103 54384711 54396114 54397785 54436980 54474018 54526149 54526578 54574047 54577215 54618864 54635859 54898581 54961113 54990756 55002486 55017399 55059921 55118052 55144227 55185585 55293441 55303632 55355454 55422465 55443366 55463280 55487745 55526550 55637793 55735635 55755747 55772757 55823007 55855797 55891752 55953840 56022204 56051700 56184264 56198532 56222055 56383635 56407323 56420724 56439639 56444868 56605293 56671821 56687220 56714592 56762799 56844414 56867526 56937315 57041106 57119916 57248325 57331905 57370308 57426795 57430518 57433047 57581919 57648276 57670254 57685986 57695196 57704628 57827754 57892995 57967926 57977379 58013700 58077978 58106796 58233828 58263339 58294272 58303356 58330077 58347060 58361331 58406220 58426233 58427844 58449162 58469034 58574217 58612884 58626234 58696500 58727904 58787271 58799478 58808826 58849362 58891002 58904472 58909557 58982721 59008218 59073057 59194323 59246646 59379303 59382996 59390001 59407032 59450433 59487663 59568357 59572308 59579388 59589948 59595948 59651424 59671500 59769024 59833224 59854032 59933754 59970747 60061857 60131166 60162666 60199323 60334446 60497433 60527508 60640446 60710298 60772269 60791016 60791763 60801453 60829476 60853587 61001082 61077756 61100748 61179933 61276218 61281204 61290597 61361844 61432482 61471407 61478904 61481622 61519929 61540830 61542336 61561407 61657731 61723635 61769322 61790154 61809318 61822095 61827177 61934556 61936299 61950912 61983201 62064291 62122320 62192649 62246367 62246703 62429142 62430672 62476350 62484888 62523222 62693316 62758404 62774130 62933592 63076011 63083169 63207321 63234897 63297630 63330024 63383667 63432690 63522336 63536982 63619794 63621045 63717834 63800709 63882672 63940212 63981948 64004028 64166715 64176405 64180269 64252920 64256751 64344687 64410513 64604100 64610625 64640121 64667616 64777128 64840314 64859109 64888377 64911717 65033553 65157384 65189859 65220315 65281461 65284332 65358339 65386320 65417301 65427111 65459166 65531409 65567208 65657310 65766012 65844234 65867490 66021975 66037038 66040719 66081843 66114477 66115752 66127512 66142311 66157869 66169185 66178017 66269412 66287397 66382464 66501135 66509520 66608961 66662064 66681135 66787077 66852291 66903963 67135269 67181160 67205307 67253886 67305633 67360035 67382280 67458090 67470402 67478355 67520145 67564770 67621644 67661316 67822965 67842444 67842591 67919763 67930515 67957041 68028636 68033556 68116470 68161755 68182206 68203095 68237115 69301650 71236728 71978445 71994045 74134488 74643216 74650032 74713728 74717226 74753760 74772066 74828727 74895447 74950152 74968623 74972460 75065256 75171804 75205308 75245805 75295113 75302484 75364059 75644811 75697695 75719718 75876030 75977604 75998310 76010628 76025055 76088670 76270296 76478232 76750110 76890591 76971999 77068689 77198139 77259399 77272779 77347545 77717787 77938806 78202500 78212136 78435468 78889542 78947658 78990159 79224816 79357362 79368873 79902759 80080437 80194860 80299659 80306547 80387838 80561265 80786586 80923131 80993496 81004107 81024651 81024762 81088257 81095964 81170610 81245379 81277266 81331959 81428790 81448956 81452493 81465534 81549942 82244130 82281258 82298214 82453536 82708194 82780320 82897437 82962873 82975854 82979220 83037204 83044689 83050491 83083809 83168280 83197551 83314737 83320158 83353620 83356419 83404713 83405898 83445660 83452806 83460846 83549820 83787513 83788818 83811168 83842269 83865018 83868399 83914683 83916594 84018207 84029562 84043530 84047559 84078600 84088335 84110052 84149655 84221460 84435606 84441000 84615627 84731874 84739929 84914610 84926853 84968712 85082523 85084815 85096014 85119834 85135275 85197462 85220796 85263006 85406229 85453944

Since only 600 out of 3000 are affected we go aheand and revode the cases form the online order table.

In [70]:
-- Delete rows with unmatched customer numbers
DELETE FROM WOODCORP_ETC_ONLINE
WHERE CUSTOMER_NUMBER_ONLINE IN (SELECT CUSTOMER_NUMBER_ONLINE 
                                 FROM WOODCORP_ETC_ONLINE
                                 WHERE CUSTOMER_NUMBER_ONLINE NOT IN (SELECT CUSTOMER_NO FROM WOODCORP_ETC_CUSTOMERDATA));


(1453 rows affected)

Total execution time: 00:00:00.809

In [71]:
-- Drop existing primary key constraint if any for WOODCORP_ETC_ONLINE
IF EXISTS (SELECT * FROM information_schema.table_constraints
           WHERE table_name = 'WOODCORP_ETC_ONLINE'
           AND constraint_type = 'PRIMARY KEY')
BEGIN
    DECLARE @sql1 NVARCHAR(MAX)
    SET @sql1 = (SELECT 'ALTER TABLE WOODCORP_ETC_ONLINE DROP CONSTRAINT ' + constraint_name
                FROM information_schema.table_constraints
                WHERE table_name = 'WOODCORP_ETC_ONLINE'
                AND constraint_type = 'PRIMARY KEY')
    EXEC sp_executesql @sql1
END

-- Add primary key constraint on ORDER_NO for WOODCORP_ETC_ONLINE
ALTER TABLE WOODCORP_ETC_ONLINE
ADD CONSTRAINT WOODCORP_ETC_ONLINE_PK PRIMARY KEY (ORDER_NO);

-- Drop existing foreign key constraint if any for WOODCORP_ETC_ONLINE
IF EXISTS (SELECT * FROM information_schema.table_constraints
           WHERE table_name = 'WOODCORP_ETC_ONLINE'
           AND constraint_type = 'FOREIGN KEY')
BEGIN
    DECLARE @sql2 NVARCHAR(MAX)
    SET @sql2 = (SELECT 'ALTER TABLE WOODCORP_ETC_ONLINE DROP CONSTRAINT ' + constraint_name
                FROM information_schema.table_constraints
                WHERE table_name = 'WOODCORP_ETC_ONLINE'
                AND constraint_type = 'FOREIGN KEY')
    EXEC sp_executesql @sql2
END

-- Add foreign key constraint on CUSTOMER_NUMBER_ONLINE for WOODCORP_ETC_ONLINE
ALTER TABLE WOODCORP_ETC_ONLINE
ADD CONSTRAINT WOODCORP_ETC_ONLINE_FK FOREIGN KEY (CUSTOMER_NUMBER_ONLINE)
REFERENCES WOODCORP_ETC_CUSTOMERDATA (CUSTOMER_NO);



Commands completed successfully.

Total execution time: 00:00:01.060

### **Transformations Offline Table**

In [72]:
-- Check data types
SELECT COLUMN_NAME, DATA_TYPE
FROM INFORMATION_SCHEMA.COLUMNS
WHERE TABLE_NAME = 'WOODCORP_ETC_OFFLINE';

(8 rows affected)

Total execution time: 00:00:00.543

COLUMN_NAME DATA_TYPE ORDER_NO float CUSTOMER_NUMBER_OFFLINE int ORDER_VALUE_OFFLINE float DISCOUNT_VALUE_OFFLINE float NUMBER_ITEMS_OFFLINE smallint NUMBER_DISCOUNT_ITEMS_OFFLINE tinyint HIGH_QUALITY_WOOD tinyint LOW_QUALITY_WOOD tinyint

All is good:  
\- SmallInt for values above 255, TinyInt for values below 255, Float is fine.  
<span style="font-family: -apple-system, BlinkMacSystemFont, sans-serif; color: var(--vscode-foreground);">- Just need to convert ORDER_NO to Int.</span>

In [80]:
-- Transform ORDER_NO to a smaller range
UPDATE dbo.WOODCORP_ETC_OFFLINE
SET ORDER_NO = ORDER_NO / 1000000000; -- Adjust the divisor based on the number of extraneous zeros

(134758 rows affected)

Total execution time: 00:00:01.441

In [85]:
-- Change the data type of ORDER_NO to INT NOT NULL (SQL server seems to find null values while there aren't)
ALTER TABLE dbo.WOODCORP_ETC_OFFLINE
ALTER COLUMN ORDER_NO INT NOT NULL;


Commands completed successfully.

Total execution time: 00:00:01.281

In [1]:
-- Identify duplicate ORDER_NO values as we seem to encounter this issue when creating the keys
SELECT ORDER_NO, COUNT(*)
FROM dbo.WOODCORP_ETC_OFFLINE
GROUP BY ORDER_NO
HAVING COUNT(*) > 1;


(8178 rows affected)

Displaying Top 5000 rows.

Total execution time: 00:00:01.459

ORDER_NO (No column name) 20200 71 293050 11 213750 24 20710 132 21830 32 217560 10 188520 9 264520 5 218240 4 22580 60 20980 116 181140 2 208820 24 24790 45 286520 10 283120 19 18770 72 212800 15 20810 68 219260 2 291690 4 185120 2 190730 6 248810 24 197020 4 185900 18 28700 70 197890 13 19690 148 272630 4 27680 57 20930 71 19790 125 27950 47 227420 13 203380 7 183860 9 196240 3 20590 84 25810 29 204400 27 194980 11 186990 8 217730 3 19280 67 185800 5 25980 21 21610 78 291280 18 27850 22 225380 6 184100 10 28460 35 267750 3 292300 8 1940 67 19960 65 18550 98 210520 17 292030 14 204910 16 285670 7 223850 6 185970 22 27440 76 21660 56 270640 14 22680 103 2890 54 217900 11 295770 14 29040 56 220910 16 22340 62 286690 18 285500 17 192770 19 21950 45 205420 6 278800 10 25300 36 199910 9 213770 20 203890 20 249320 34 26080 11 29550 55 215280 6 18600 90 265300 34 27730 64 21560 70 183520 20 210860 13 270300 7 194010 13 21320 74 28530 53 263790 11 248010 3 210180 14 283630 7 294920 5 184880 26 193790 8 2720 37 292710 27 208890 7 206100 16 24960 38 269960 25 287930 26 191070 6 28360 61 25060 93 269910 5 219600 2 287830 12 223000 12 182400 5 185750 3 206440 8 181380 7 181770 2 28870 74 2040 109 282050 13 20690 81 262920 5 19910 62 25690 40 257650 9 19570 85 250050 19 198890 7 292200 10 196630 15 195610 4 204010 13 251700 11 18940 43 191870 8 20300 32 276420 3 190560 3 296790 15 19180 62 281810 5 254030 10 193010 14 200760 22 286930 14 279090 4 209910 16 294630 12 291790 22 289820 4 202020 19 211880 7 226910 7 202750 18 204790 14 221640 3 185630 15 287710 11 254660 5 29720 97 210760 10 220790 16 254810 28 2550 60 296670 8 190900 10 212900 10 197990 21 209160 17 294070 17 293320 12 290670 4 184780 12 215910 11 197770 17 250000 6 207120 17 216640 7 284770 2 211780 8 256070 21 193910 11 182500 15 272340 5 288900 12 201630 7 270690 4 205590 11 267580 18 261800 10 26930 16 27340 27 288800 8 224020 2 295650 16 25570 27 199690 13 203770 15 212630 7 209890 13 190220 17 198280 22 253060 5 247670 3 251070 14 291960 25 263940 17 219890 7 221400 10 179730 13 272050 7 221930 3 199400 15 295700 19 209670 13 248690 8 272000 8 249830 3 225280 8 190050 20 288850 16 277680 12 208140 6 204740 10 225790 9 183760 7 222830 4 206610 17 276320 6 214260 4 256120 17 273700 6 282950 15 197360 9 280790 10 291180 12 293900 16 224650 3 2670 19 186870 6 249420 15 287030 16 223050 12 274670 6 198790 7 252720 12 184030 23 278020 11 262820 7 198040 16 180360 9 297060 14 181990 12 226690 12 275010 3 225670 7 210910 14 219990 5 260780 4 207630 8 209400 15 210690 14 194930 6 285160 4 189030 8 258500 5 213920 8 247910 3 290770 10 237690 3 253010 20 268670 11 200030 13 280670 6 257140 7 187670 25 226010 3 207020 6 205930 27 26660 15 276710 4 191140 9 213140 6 277730 6 295940 26 207800 9 249030 12 289920 6 192940 8 270810 14 26950 12 217660 3 297640 6 288560 4 287440 5 283970 6 290940 11 216880 4 294970 13 284140 8 275400 5 274790 4 290090 5 222320 7 26710 10 264280 17 216930 20 224770 5 2513700 4 187160 6 260540 10 250680 8 280690 8 297810 5 294680 22 259010 3 258670 5 23990 19 188010 2 191750 17 297690 4 205520 14 202630 7 256800 9 187500 18 188910 16 210010 3 225890 21 296960 19 185850 3 180020 4 193520 3 206780 6 195900 7 188690 5 181480 5 189640 10 282540 6 211030 12 192600 6 193280 11 196920 16 255050 3 206950 7 181650 5 281180 8 257820 2 201050 5 201780 2 197870 6 251870 7 186480 12 205030 11 290890 3 197650 17 276830 18 187770 7 26320 9 265930 6 293830 3 214770 2 255440 6 179780 4 26590 7 259420 3 227030 2 283050 14 182740 10 251020 6 183010 8 274040 8 213240 5 215180 17 284820 3 295090 6 214500 3 279040 5 292930 17 204060 4 247890 7 254420 9 210810 4 272920 3 226400 8 180630 10 188860 16 195660 12 214670 9 214890 3 273020 11 187840 8 273650 10 208480 4 222030 8 192890 11 26830 12 182910 2 286810 3 253400 2 201000 14 199640 7 291910 11 225040 3 280910 2 186770 7 259930 6 206050 4 255950 4 183810 4 228050 9 209790 4 215010 7 294580 17 200660 3 212050 3 288220 5 276930 

In [2]:
-- Add a surrogate key column
ALTER TABLE WOODCORP_ETC_OFFLINE ADD ID INT IDENTITY(1,1) PRIMARY KEY;


Commands completed successfully.

Total execution time: 00:00:04.581

A surrogate key is an artificial or synthetic key that is used to uniquely identify each record in a table. It is useful when there are no natural primary keys or when natural primary keys have issues such as duplicates or are too complex. By adding a surrogate key column (ID), we ensure each record has a unique identifier.

In [3]:
-- Drop existing primary key constraint if any
IF EXISTS (SELECT * FROM information_schema.table_constraints
           WHERE table_name = 'WOODCORP_ETC_OFFLINE'
           AND constraint_type = 'PRIMARY KEY')
BEGIN
    DECLARE @sql3 NVARCHAR(MAX)
    SET @sql3 = (SELECT 'ALTER TABLE WOODCORP_ETC_OFFLINE DROP CONSTRAINT ' + constraint_name
                FROM information_schema.table_constraints
                WHERE table_name = 'WOODCORP_ETC_OFFLINE'
                AND constraint_type = 'PRIMARY KEY')
    EXEC sp_executesql @sql3
END

-- Add primary key constraint on ID
ALTER TABLE WOODCORP_ETC_OFFLINE
ADD CONSTRAINT WOODCORP_ETC_OFFLINE_PK PRIMARY KEY (ID);

Commands completed successfully.

Total execution time: 00:00:01.235

In [6]:
-- Identify conflicting records
SELECT CUSTOMER_NUMBER_OFFLINE
FROM WOODCORP_ETC_OFFLINE
WHERE CUSTOMER_NUMBER_OFFLINE NOT IN (SELECT CUSTOMER_NO FROM WOODCORP_ETC_CUSTOMERDATA);


(57189 rows affected)

Displaying Top 5000 rows.

Total execution time: 00:00:01.488

CUSTOMER_NUMBER_OFFLINE 70589919 54846504 58005978 83858295 60618204 62728416 54353511 87649491 67837128 64127529 63426492 81909258 62146077 59122980 60012066 56701818 79730307 56056146 64111308 60560532 57915834 87582270 67311468 55859565 61686621 55225839 65498643 63838086 55888383 62992623 83057304 65734605 62706642 56557281 85597455 62412357 61813572 62196069 64801536 62062353 84002811 63904335 58891002 60163542 61352133 81988515 61681890 62759532 60438996 83521368 59003586 64267848 86024859 55227942 57533112 55837932 88395984 84107781 67671990 86252826 62620065 59055285 87377181 59861550 64215795 79355916 61596240 59698764 56971674 87582270 62385480 62815599 65473086 88401720 88831116 86436666 60139140 80908323 67750968 57101250 78842331 67569648 67048749 58953267 87187104 54362916 88763520 64864863 84818379 85521267 58979358 87151737 62243589 74866674 56461011 61178145 88868919 60034989 59543067 54996747 82133883 66101376 57311730 77731176 84422631 85262631 74370678 87345132 80162694 79824528 62945910 62170392 62652657 87678516 55453518 65784204 63780660 64508847 55140282 87864615 56516994 62954757 87312897 62962902 63784578 87071355 56331690 63840402 81940866 83163912 66054252 75626364 62452032 59446881 67671990 62886033 86512071 63655518 88595823 64236789 58140642 84879174 55153839 66216156 86625765 55922724 60285249 87972441 80654583 87648840 59765919 76452390 65553699 59512878 66535944 81330405 57964989 88591161 55647126 74966460 87069384 57702591 59404587 58049565 65784204 88010616 77704797 65058189 88318671 74666316 60187551 62078349 87941787 88969173 62966769 65633358 60045684 89005902 59042361 58790103 60915369 65192097 55538007 78918618 63511908 60807996 64412559 83945535 61456263 55488081 55468263 62281197 59108385 62107797 84078255 57600681 61576071 65915295 58597959 57514479 84360606 55151175 58591347 84339120 77296416 63667455 56453208 86447133 87244221 86110473 89323068 64591740 86837250 55255671 54649869 79144680 63208518 57183915 58854141 58542279 84479493 81789381 88774596 64613157 61357914 55293318 76779906 63776343 55275795 63143070 75769440 63107211 64977159 60967812 68263968 59911482 62398464 64154496 55318428 88401519 56459967 64676259 64671597 75713019 86403510 85084896 60752160 57149862 55850568 86000196 81321729 56245221 58693884 57711897 59316207 66907824 63824487 59161698 55488081 53802729 64887522 55239255 83876124 63509877 68225895 62342841 61014204 84469206 64236789 62157366 56708235 88586190 85541712 86648517 80334777 64349376 67563915 68233134 63370653 68041842 65434176 56506101 61531938 57115323 86447133 62967963 63613041 74493426 53824257 64804653 76869498 76581207 86896248 76062243 58745577 62623971 65287470 64816908 57238107 83202660 60026349 57008424 61322433 77097693 74970726 87071106 86312352 58495860 58520988 87994152 87940548 62351199 65386683 56406912 66081843 54952644 81598419 85297518 57815835 66939135 84677349 85144173 59446881 58963086 60193203 87342660 64448025 81028902 60804045 75549591 81597903 60238929 86828547 63944880 57854349 67046820 60987852 54975015 63015879 89246688 86395614 84120909 63016704 65911932 85618188 88318296 57241095 66151125 71302041 81565146 60308229 65089104 88197771 86828547 64214508 60798993 57226416 57425121 66287211 59162097 88197771 54592926 61280811 62113035 81797544 59981865 55351221 75972996 63644943 63697734 63318732 54899337 68306985 56502285 74934285 61129491 63715860 68036823 63316506 55523262 56718654 85757850 64131906 88550331 68145687 64840314 87626670 61005702 56214546 64324752 59021628 57947592 75734379 64833255 61441041 58959702 67432590 56612817 63934623 75324585 67754598 86033226 65604714 55110567 83094039 56290134 65551722 83861808 83575563 75108927 63811824 85183779 64924674 59822913 65183253 75870306 67164621 57252756 82149885 88521048 55826814 62578884 57119916 62663880 61697265 81789381 75065775 82138257 67616529 81737325 84498738 75626364 62615181 76998558 74866674 64516548 89214534 61167789 59436459 87877203 62414967 5499390

In [7]:
-- Remove invalid records
DELETE FROM WOODCORP_ETC_OFFLINE
WHERE CUSTOMER_NUMBER_OFFLINE NOT IN (SELECT CUSTOMER_NO FROM WOODCORP_ETC_CUSTOMERDATA);


(57189 rows affected)

Total execution time: 00:00:01.113

Repeating the approach followed for the ONLINE dataset we removed the orders of customers not present in our customer database. (

In [8]:
-- Drop existing foreign key constraint if any
IF EXISTS (SELECT * FROM information_schema.table_constraints
           WHERE table_name = 'WOODCORP_ETC_OFFLINE'
           AND constraint_type = 'FOREIGN KEY')
BEGIN
    DECLARE @sql4 NVARCHAR(MAX)
    SET @sql4 = (SELECT 'ALTER TABLE WOODCORP_ETC_OFFLINE DROP CONSTRAINT ' + constraint_name
                FROM information_schema.table_constraints
                WHERE table_name = 'WOODCORP_ETC_OFFLINE'
                AND constraint_type = 'FOREIGN KEY')
    EXEC sp_executesql @sql4
END

-- Add foreign key constraint on CUSTOMER_NUMBER_OFFLINE
ALTER TABLE WOODCORP_ETC_OFFLINE
ADD CONSTRAINT WOODCORP_ETC_OFFLINE_FK FOREIGN KEY (CUSTOMER_NUMBER_OFFLINE)
REFERENCES WOODCORP_ETC_CUSTOMERDATA (CUSTOMER_NO);

Commands completed successfully.

Total execution time: 00:00:00.393

### **Null Values Handdling**

In [7]:
-- Check for null values in key columns
SELECT COUNT(*) AS NullCustomerNo
FROM dbo.WOODCORP_ETC_CUSTOMERDATA
WHERE CUSTOMER_NO IS NULL;

(1 row affected)

Total execution time: 00:00:00.394

NullCustomerNo 0

In [2]:

SELECT COUNT(*) AS NullOrderNoOnline
FROM dbo.WOODCORP_ETC_ONLINE
WHERE ORDER_NO IS NULL;


(1 row affected)

Total execution time: 00:00:00.465

NullOrderNoOnline 0

In [3]:
SELECT COUNT(*) AS NullCustomerNumberOnline
FROM dbo.WOODCORP_ETC_ONLINE
WHERE CUSTOMER_NUMBER_ONLINE IS NULL;

(1 row affected)

Total execution time: 00:00:00.315

NullCustomerNumberOnline 0

In [4]:
SELECT COUNT(*) AS NullOrderNoOffline
FROM dbo.WOODCORP_ETC_OFFLINE
WHERE ORDER_NO IS NULL;

(1 row affected)

Total execution time: 00:00:00.292

NullOrderNoOffline 0

In [5]:
SELECT COUNT(*) AS NullCustomerNumberOffline
FROM dbo.WOODCORP_ETC_OFFLINE
WHERE CUSTOMER_NUMBER_OFFLINE IS NULL;

(1 row affected)

Total execution time: 00:00:00.341

NullCustomerNumberOffline 0

There aren't any null values.

## **2) Exploratory Data Analysis**

### **Basic Questions**

In [11]:
-- 1) How many unique customers are there?
SELECT COUNT(DISTINCT CUSTOMER_NO) AS UniqueCustomers 
FROM dbo.WOODCORP_ETC_CUSTOMERDATA;


(1 row affected)

Total execution time: 00:00:00.395

UniqueCustomers 13221

In [4]:
-- 2) What is the average order value online?
SELECT ROUND(AVG(ORDER_VALUE_ONLINE), 0) AS AverageOrderValueOnline 
FROM dbo.WOODCORP_ETC_ONLINE;


(1 row affected)

Total execution time: 00:00:00.279

AverageOrderValueOnline 17219

In [10]:
-- Check for offline
SELECT ROUND(AVG(ORDER_VALUE_OFFLINE), 0) AS AverageOrderValueOnline 
FROM dbo.WOODCORP_ETC_OFFLINE;

(1 row affected)

Total execution time: 00:00:00.247

AverageOrderValueOnline 4180

WoodCorp's online orders total 835,606 items, while offline orders total 2,456,720 items. Additionally, the average value of online orders is $17,219, compared to $4,180 for offline orders. This suggests that while the volume of transactions is higher offline, the value per transaction is significantly higher online, indicating potential growth opportunities for high-value sales through online channels while leveraging the high volume of offline transactions.

In [15]:
-- 3) What is our most expensive order?
SELECT ROUND(MAX(ORDER_VALUE_ONLINE), 0) AS MostExpensiveOrder 
FROM dbo.WOODCORP_ETC_ONLINE;


(1 row affected)

Total execution time: 00:00:00.067

MostExpensiveOrder 208788

In [16]:
-- 4) Who (customer number) ordered our most expensive order?
SELECT CUSTOMER_NUMBER_ONLINE 
FROM dbo.WOODCORP_ETC_ONLINE 
WHERE ORDER_VALUE_ONLINE = (SELECT MAX(ORDER_VALUE_ONLINE) FROM dbo.WOODCORP_ETC_ONLINE);


(1 row affected)

Total execution time: 00:00:00.038

CUSTOMER_NUMBER_ONLINE 86040033

In [17]:
-- 5) How many customers are part of the loyalty program?
SELECT COUNT(*) AS LoyaltyProgramMembers 
FROM dbo.WOODCORP_ETC_CUSTOMERDATA 
WHERE LOYALTY_PROGRAM = 1;


(1 row affected)

Total execution time: 00:00:00.040

LoyaltyProgramMembers 4604

34.8% of the 13,221 unique customers are in the loyalty program. This is good but we could consider increasing the enrolment in the program with enhance benefits, targeted marketing, and referral incentives to boost loyalty program participation.

In [18]:
-- 6) How many orders include high quality wood?
SELECT COUNT(*) AS HighQualityOrders 
FROM dbo.WOODCORP_ETC_OFFLINE 
WHERE HIGH_QUALITY_WOOD = 1;


(1 row affected)

Total execution time: 00:00:00.241

HighQualityOrders 15368

Out of 77,569 offline orders, 15,368 include high-quality wood. This represents approximately 19.82% of all offline orders. This indicates a significant demand for high-quality wood, suggesting the company should continue to focus on and possibly expand its high-quality wood product offerings to meet customer demand. Of course also depending on what is the benefit made out of high quality versus low.

In [20]:
-- 7) What percentage of customers is older than 55 years?
SELECT ROUND((COUNT(*) * 100.0 / (SELECT COUNT(*) FROM dbo.WOODCORP_ETC_CUSTOMERDATA)), 0) AS PercentageOlderThan55
FROM dbo.WOODCORP_ETC_CUSTOMERDATA 
WHERE AGE = 4;


(1 row affected)

Total execution time: 00:00:00.058

PercentageOlderThan55 15.000000000000

15% of WoodCorp's customers are older than 55 years, indicating they are underrepresented compared to an even distribution of 25%.  
Strategic Options:  
\- **Increase Efforts for Older Customers**: Target marketing and products to attract more older customers, leveraging their potential value.  
\- **Focus on Current Strengths**: Continue to engage and satisfy demographics with higher proportions to maximize returns.  

—\> Balanced approach.

In [5]:
-- 8) Of what customer type do we have the least customers apart from customer type 0?
SELECT TOP 1 
    CASE 
        WHEN CUSTOMER_TYPE = 1 THEN 'Transporter'
        WHEN CUSTOMER_TYPE = 2 THEN 'Factory'
        WHEN CUSTOMER_TYPE = 3 THEN 'Trader'
        WHEN CUSTOMER_TYPE = 4 THEN 'Private'
        ELSE 'Unknown'
    END AS CUSTOMER_TYPE,
    COUNT(*) AS Count
FROM dbo.WOODCORP_ETC_CUSTOMERDATA
WHERE CUSTOMER_TYPE <> 0
GROUP BY CUSTOMER_TYPE
ORDER BY Count ASC;


(1 row affected)

Total execution time: 00:00:00.042

CUSTOMER_TYPE Count Private 1026

To improve the market penetration and increase the customer base in the "Private" segment, WoodCorp should consider targeted marketing strategies and promotions. This might include personalized offers, loyalty programs, and tailored communications to attract more private customers.

In [25]:
-- 9) How many items have been ordered online?
SELECT SUM(NUMBER_ITEMS_ONLINE) AS TotalItemsOrderedOnline 
FROM dbo.WOODCORP_ETC_ONLINE;


(1 row affected)

Total execution time: 00:00:00.036

TotalItemsOrderedOnline 835606

WoodCorp's online orders total 835,606 items, while offline orders total 2,456,720 items. Additionally, the average value of online orders is $17,219, compared to $4,180 for offline orders. This suggests that while the volume of transactions is higher offline, the value per transaction is significantly higher online, indicating potential growth opportunities for high-value sales through online channels while leveraging the high volume of offline transactions.

### **Intermediate Questions**

In [26]:
-- 1) How many women are part of the loyalty program?
SELECT COUNT(*) AS WomenInLoyaltyProgram
FROM dbo.WOODCORP_ETC_CUSTOMERDATA
WHERE GENDER = 1  -- Female
AND LOYALTY_PROGRAM = 1;  -- In loyalty program



(1 row affected)

Total execution time: 00:00:00.050

WomenInLoyaltyProgram 1449

Additional investigations to see how much it represents and if men observe different proportions:

In [3]:
-- Calculate the ratio of women in the loyalty program
SELECT 
    (SELECT COUNT(*) 
     FROM dbo.WOODCORP_ETC_CUSTOMERDATA
     WHERE GENDER = 1 
     AND LOYALTY_PROGRAM = 1) * 1.0 / 
    (SELECT COUNT(*) 
     FROM dbo.WOODCORP_ETC_CUSTOMERDATA
     WHERE GENDER = 1) AS RatioWomenInLoyaltyProgram;


(1 row affected)

Total execution time: 00:00:00.038

RatioWomenInLoyaltyProgram 0.343364928909

In [4]:
-- Calculate the ratio of men in the loyalty program
SELECT 
    (SELECT COUNT(*) 
     FROM dbo.WOODCORP_ETC_CUSTOMERDATA
     WHERE GENDER = 0 
     AND LOYALTY_PROGRAM = 1) * 1.0 / 
    (SELECT COUNT(*) 
     FROM dbo.WOODCORP_ETC_CUSTOMERDATA
     WHERE GENDER = 0) AS RatioMenInLoyaltyProgram;


(1 row affected)

Total execution time: 00:00:00.053

RatioMenInLoyaltyProgram 0.350516609265

Out of the total female customers, approximately 34.34% are enrolled in the loyalty program. In comparison, around 35.05% of male customers are part of the loyalty program. This indicates a similar level of engagement with the loyalty program between male and female customers, suggesting that promotional efforts are reaching both demographics effectively and should only be accentuated but no change needed in the format.

In [6]:
-- 2) What customer type is relatively speaking responsible for most online orders?
SELECT 
    CASE 
        WHEN CUSTOMER_TYPE = 1 THEN 'Transporter'
        WHEN CUSTOMER_TYPE = 2 THEN 'Factory'
        WHEN CUSTOMER_TYPE = 3 THEN 'Trader'
        WHEN CUSTOMER_TYPE = 4 THEN 'Private'
        ELSE 'Unknown'
    END AS CUSTOMER_TYPE,
    COUNT(*) AS OrderCount
FROM dbo.WOODCORP_ETC_ONLINE AS O
JOIN dbo.WOODCORP_ETC_CUSTOMERDATA AS C
ON O.CUSTOMER_NUMBER_ONLINE = C.CUSTOMER_NO
GROUP BY CUSTOMER_TYPE
ORDER BY OrderCount DESC;


(5 rows affected)

Total execution time: 00:00:00.055

CUSTOMER_TYPE OrderCount Factory 2874 Transporter 2869 Trader 825 Private 348 Unknown 111

This insight suggests that efforts to optimize online sales processes should prioritize factories and transporters due to their high order volume.

In [7]:
-- 3) What customer type is on average the youngest?
SELECT 
    CASE 
        WHEN CUSTOMER_TYPE = 1 THEN 'Transporter'
        WHEN CUSTOMER_TYPE = 2 THEN 'Factory'
        WHEN CUSTOMER_TYPE = 3 THEN 'Trader'
        WHEN CUSTOMER_TYPE = 4 THEN 'Private'
        ELSE 'Unknown'
    END AS CUSTOMER_TYPE,
    ROUND(AVG(CAST(AGE AS FLOAT)), 5) AS AvgAge
FROM dbo.WOODCORP_ETC_CUSTOMERDATA
GROUP BY CUSTOMER_TYPE
ORDER BY AvgAge ASC;


(5 rows affected)

Total execution time: 00:00:00.043

CUSTOMER_TYPE AvgAge Unknown 2,21632 Private 2,21637 Trader 2,21883 Factory 2,34597 Transporter 2,76401

Understanding the age demographics can help in tailoring marketing strategies to different customer types, with a focus on engaging younger customers more effectively.

In [30]:
-- 4) What age group orders the most online items?
SELECT AGE_GROUP, SUM(NUMBER_ITEMS_ONLINE) AS TotalItemsOrdered
FROM (
    SELECT CASE
        WHEN AGE = 1 THEN '0-18'
        WHEN AGE = 2 THEN '19-35'
        WHEN AGE = 3 THEN '36-55'
        WHEN AGE = 4 THEN '55+'
        WHEN AGE = 0  THEN  'NA' END AS AGE_GROUP,
        NUMBER_ITEMS_ONLINE
    FROM dbo.WOODCORP_ETC_ONLINE AS O
    JOIN dbo.WOODCORP_ETC_CUSTOMERDATA AS C
    ON O.CUSTOMER_NUMBER_ONLINE = C.CUSTOMER_NO
) AS AgeGroupItems
GROUP BY AGE_GROUP
ORDER BY TotalItemsOrdered DESC;


(4 rows affected)

Total execution time: 00:00:00.071

AGE_GROUP TotalItemsOrdered 19-35 423244 36-55 333158 55+ 66528 0-18 12676

This insight suggests that marketing and sales strategies should focus on the 19-35 and 36-55 age groups to maximize online sales

In [32]:
-- 5) Is the value of an offline order on average higher when high-quality wood is sold?
SELECT
    CASE WHEN HIGH_QUALITY_WOOD > 0 THEN 'With High-Quality Wood' ELSE 'Without High-Quality Wood' END AS Quality,
    ROUND(AVG(ORDER_VALUE_OFFLINE), 0) AS AvgOrderValue
FROM dbo.WOODCORP_ETC_OFFLINE
GROUP BY CASE WHEN HIGH_QUALITY_WOOD > 0 THEN 'With High-Quality Wood' ELSE 'Without High-Quality Wood' END;


(2 rows affected)

Total execution time: 00:00:00.076

Quality AvgOrderValue With High-Quality Wood 7008 Without High-Quality Wood 3481

The average value of offline orders containing high-quality wood is $7,008, compared to $3,481 for those without. This indicates that high-quality wood significantly increases order value. Further analysis could determine if the higher order values translate to higher overall profitability.

In [36]:
-- 6) Is the average value of an online order higher when more than 5 items are bought in discount?
SELECT
    CASE WHEN NUMBER_DISCOUNTED_ITEMS_ONLINE > 5 THEN 'More than 5 Discounted Items' ELSE '5 or Less Discounted Items' END AS DiscountCategory,
    ROUND(AVG(ORDER_VALUE_ONLINE), 0) AS AvgOrderValue
FROM dbo.WOODCORP_ETC_ONLINE
GROUP BY CASE WHEN NUMBER_DISCOUNTED_ITEMS_ONLINE > 5 THEN 'More than 5 Discounted Items' ELSE '5 or Less Discounted Items' END;


(2 rows affected)

Total execution time: 00:00:00.064

DiscountCategory AvgOrderValue More than 5 Discounted Items 18572 5 or Less Discounted Items 13454

Orders with more than 5 discounted items have a higher average value of $18,572 compared to $13,454 for those with 5 or fewer discounted items. This suggests that offering bulk discounts can significantly increase the order value.

In [33]:
-- 7) How many customers placed more than 3 orders in total? Distinguish between online and offline.
SELECT
    CUSTOMER_NO,
    SUM(CASE WHEN Source = 'Online' THEN 1 ELSE 0 END) AS OnlineOrders,
    SUM(CASE WHEN Source = 'Offline' THEN 1 ELSE 0 END) AS OfflineOrders,
    SUM(OrderCount) AS TotalOrders
FROM (
    SELECT CUSTOMER_NUMBER_ONLINE AS CUSTOMER_NO, COUNT(*) AS OrderCount, 'Online' AS Source
    FROM dbo.WOODCORP_ETC_ONLINE
    GROUP BY CUSTOMER_NUMBER_ONLINE
    UNION ALL
    SELECT CUSTOMER_NUMBER_OFFLINE AS CUSTOMER_NO, COUNT(*) AS OrderCount, 'Offline' AS Source
    FROM dbo.WOODCORP_ETC_OFFLINE
    GROUP BY CUSTOMER_NUMBER_OFFLINE
) AS AllOrders
GROUP BY CUSTOMER_NO
HAVING SUM(OrderCount) > 3;


(8615 rows affected)

Displaying Top 5000 rows.

Total execution time: 00:00:00.236

CUSTOMER_NO OnlineOrders OfflineOrders TotalOrders 57689514 1 1 5 78510288 1 1 7 58986849 0 1 15 88641975 0 1 14 64614207 0 1 8 81059568 0 1 15 88951077 0 1 4 62647728 0 1 4 57990126 1 1 7 62232117 0 1 14 66602103 0 1 4 74822373 0 1 5 88450620 0 1 7 61779990 0 1 5 61846728 1 1 9 55687701 1 1 13 86776263 0 1 10 56440632 0 1 33 75968001 1 1 9 54683739 0 1 6 54456858 0 1 16 56247465 1 0 6 61761444 0 1 10 57279837 0 1 10 66847146 0 1 9 83658291 0 1 5 86392056 0 1 31 63190635 0 1 6 58110936 0 1 4 59199744 0 1 7 57213222 1 0 5 64672332 0 1 9 77010360 0 1 5 63485505 0 1 10 88504308 0 1 13 55853271 1 1 5 63268926 1 1 6 54692037 1 1 6 57631335 0 1 11 57794649 0 1 9 55791339 0 1 11 66561081 0 1 11 62233560 0 1 16 55466193 0 1 9 63043419 0 1 6 84619365 0 1 6 84078270 1 1 5 57419001 0 1 12 60575055 0 1 6 54035697 0 1 14 83609907 0 1 4 81219888 0 1 4 62516931 0 1 5 54284364 1 1 8 55764495 1 0 5 88344426 0 1 9 87456891 0 1 19 63604434 0 1 10 89065638 0 1 5 68058384 1 1 9 83319351 0 1 9 58896384 0 1 9 65948688 0 1 9 55835355 0 1 4 75202650 1 1 8 77904696 0 1 7 59769249 0 1 14 84310455 0 1 6 63071952 1 1 7 78490245 1 1 5 61218099 0 1 6 68273397 0 1 7 63064275 1 1 8 76445913 1 0 4 88563684 0 1 9 88190838 0 1 6 84585774 0 1 19 62930109 0 1 10 56096073 0 1 12 84092694 1 1 9 67866153 1 1 7 67814715 0 1 8 59580450 0 1 6 77032398 0 1 11 60218007 0 1 5 60873540 0 1 7 57520389 1 1 4 78054162 0 1 13 85697565 0 1 6 54716079 0 1 6 85283949 0 1 11 57616911 0 1 8 86477340 1 1 9 56761533 1 1 10 87910023 0 1 15 78529701 0 1 21 81822609 0 1 17 59199990 0 1 15 81238680 0 1 10 58575429 0 1 17 84467160 0 1 8 54837504 0 1 4 65685474 0 1 12 85294812 0 1 10 53924568 0 1 10 81975936 0 1 22 78603993 0 1 5 55454142 0 1 8 55864263 1 1 8 68175072 0 1 13 54455799 0 1 11 62585919 0 1 7 55818813 0 1 4 83511513 1 0 6 64999098 1 0 4 61924458 1 1 8 84403914 0 1 6 54851805 0 1 12 83512011 0 1 5 64331328 0 1 7 86951628 0 1 5 58993281 0 1 8 58098630 0 1 10 55494105 0 1 7 62448444 0 1 9 55159095 1 1 5 56328819 0 1 9 57423807 1 1 7 65598879 0 1 7 79491213 0 1 14 55249683 0 1 5 64839960 0 1 7 80479077 0 1 6 75210204 0 1 13 64219329 0 1 4 58502067 0 1 9 57413628 0 1 13 89192934 0 1 9 58630431 0 1 4 60219879 0 1 14 89166774 0 1 6 59834052 0 1 13 63558921 0 1 6 65247327 0 1 11 58402176 0 1 13 84222738 0 1 7 55899474 1 0 11 55878249 1 0 5 65256009 0 1 5 57918768 0 1 21 84740982 0 1 7 87327162 0 1 7 59867889 0 1 9 81122436 0 1 6 67576911 1 1 15 88713333 0 1 10 88584348 0 1 9 58592157 1 0 6 54522597 1 1 8 74455836 1 0 4 63128955 0 1 5 58842948 1 1 5 59553168 0 1 8 63182835 0 1 9 63117903 0 1 7 88062921 0 1 4 87277527 0 1 4 67001040 0 1 7 83866251 0 1 5 55586313 0 1 13 62699358 1 1 5 54043935 1 1 13 87374922 0 1 32 62078661 0 1 11 84135081 0 1 14 66825477 0 1 6 87508083 0 1 7 59373552 0 1 21 86182785 0 1 11 60770088 0 1 17 81994920 0 1 5 85434414 0 1 6 54232296 1 1 17 61612983 1 0 4 88442691 0 1 18 66333333 0 1 4 64561518 0 1 8 65440056 0 1 4 56148204 0 1 4 86583036 0 1 7 59426994 0 1 7 57744585 0 1 13 64189545 0 1 12 87361245 0 1 4 89062392 0 1 9 64536915 0 1 4 54846123 0 1 7 60988170 0 1 5 87443280 0 1 7 79810869 0 1 9 59330163 1 1 7 54111300 1 1 10 67769385 0 1 13 87047904 1 1 7 60402369 1 0 5 83891721 1 1 5 82727493 0 1 6 87681708 0 1 8 66700620 0 1 5 53896719 0 1 11 84888573 0 1 14 66369540 0 1 13 57750948 0 1 13 81774471 0 1 15 66782088 0 1 6 80218167 0 1 5 56810169 0 1 7 56114298 0 1 24 57158283 0 1 8 88404171 0 1 17 75209013 0 1 8 62711910 0 1 6 88781193 0 1 17 54963117 0 1 5 55307745 0 1 4 59934942 0 1 8 58715775 0 1 5 57327537 0 1 5 87623520 0 1 8 64237056 0 1 10 64879356 0 1 4 62224065 1 1 6 85417494 0 1 10 81410868 0 1 22 89220093 0 1 6 68130864 0 1 4 84173418 1 1 9 65288034 0 1 6 76969905 0 1 7 77567193 0 1 4 62750682 0 1 10 61316373 0 1 4 83388705 1 1 4 80990076 0 1 17 74751888 0 1 9 55141047 0 1 4 81615954 0 1 13 81669027 0 1 10 57859023 0 1 5 53884359 0 1 6 59503098 0 1 12 57325044 1 1 6 58454805 0 1 4 80550435 0 1 6 87179445 0 1 4 61224717 1 1 4 76050660 1 1 9 54287043 1

In [8]:
-- Summ onliners
SELECT COUNT(*) AS OnlineCustomers
FROM (
    SELECT CUSTOMER_NUMBER_ONLINE AS CUSTOMER_NO, COUNT(*) AS OrderCount
    FROM dbo.WOODCORP_ETC_ONLINE
    GROUP BY CUSTOMER_NUMBER_ONLINE
    HAVING COUNT(*) > 3
) AS OnlineOrders;


(1 row affected)

Total execution time: 00:00:00.064

OnlineCustomers 1017

In [9]:
-- Sum Offliners
SELECT COUNT(*) AS OfflineCustomers
FROM (
    SELECT CUSTOMER_NUMBER_OFFLINE AS CUSTOMER_NO, COUNT(*) AS OrderCount
    FROM dbo.WOODCORP_ETC_OFFLINE
    GROUP BY CUSTOMER_NUMBER_OFFLINE
    HAVING COUNT(*) > 3
) AS OfflineOrders;


(1 row affected)

Total execution time: 00:00:00.071

OfflineCustomers 7695

A much larger number of customers (7695) place more than three orders offline compared to online (1017). This indicates that while the online channel is utilized, the offline channel is the dominant mode of transaction for repeat customers. Strategic initiatives to boost online engagement could be beneficial to balance this distribution and leverage the digital sales channel more effectively.

In [7]:
-- 8) What customer type orders low-quality wood more often offline?
SELECT 
    CASE 
        WHEN CUSTOMER_TYPE = 1 THEN 'Transporter'
        WHEN CUSTOMER_TYPE = 2 THEN 'Factory'
        WHEN CUSTOMER_TYPE = 3 THEN 'Trader'
        WHEN CUSTOMER_TYPE = 4 THEN 'Private'
        ELSE 'Unknown'
    END AS CUSTOMER_TYPE,
    COUNT(*) AS LowQualityWoodOrders
FROM dbo.WOODCORP_ETC_OFFLINE AS O
JOIN dbo.WOODCORP_ETC_CUSTOMERDATA AS C
ON O.CUSTOMER_NUMBER_OFFLINE = C.CUSTOMER_NO
WHERE LOW_QUALITY_WOOD > 0
GROUP BY CUSTOMER_TYPE
ORDER BY LowQualityWoodOrders DESC;



(5 rows affected)

Total execution time: 00:00:00.063

CUSTOMER_TYPE LowQualityWoodOrders Transporter 8819 Factory 4965 Trader 1467 Private 991 Unknown 453

Transporters order low-quality wood the most frequently offline, followed by factories. This indicates that transporters and factories might be more cost-sensitive or use low-quality wood for purposes that do not require high-quality materials. This insight can help in targeting marketing strategies and product recommendations for these customer types to potentially upsell or cross-sell higher quality products or complementary goods.

### **Complex Questions**

In [35]:
-- 1) How many customers ordered both online and offline?
SELECT COUNT(DISTINCT CUSTOMER_NO) AS CustomersBothOnlineOffline
FROM dbo.WOODCORP_ETC_CUSTOMERDATA AS C
WHERE EXISTS (
    SELECT 1
    FROM dbo.WOODCORP_ETC_ONLINE AS O
    WHERE O.CUSTOMER_NUMBER_ONLINE = C.CUSTOMER_NO 
) 
AND EXISTS (
    SELECT 1
    FROM dbo.WOODCORP_ETC_OFFLINE AS OFL
    WHERE OFL.CUSTOMER_NUMBER_OFFLINE = C.CUSTOMER_NO
);


(1 row affected)

Total execution time: 00:00:00.073

CustomersBothOnlineOffline 1312

Out of the entire customer base, 1,312 customers have placed orders both online and offline. This indicates a significant segment of customers who are comfortable with both purchasing channels, showing their flexibility and higher engagement with the company. By analyzing this group's buying behavior and preferences, the company can tailor marketing strategies to meet their needs better. For example, these customers might appreciate seamless transitions between online and offline experiences, such as online order pickups or in-store returns for online purchases.

In [38]:
-- 2) On average how many more/less items does a man order online?
WITH AverageItemsOrdered AS (
    SELECT GENDER, AVG(NUMBER_ITEMS_ONLINE) AS AvgItems
    FROM dbo.WOODCORP_ETC_CUSTOMERDATA AS C
    JOIN dbo.WOODCORP_ETC_ONLINE AS O
    ON C.CUSTOMER_NO = O.CUSTOMER_NUMBER_ONLINE
    GROUP BY GENDER
)
SELECT 
    (SELECT AvgItems FROM AverageItemsOrdered WHERE GENDER = 0) - 
    (SELECT AvgItems FROM AverageItemsOrdered WHERE GENDER = 1) AS DifferenceInAverageItems;


(1 row affected)

Total execution time: 00:00:00.062

DifferenceInAverageItems 1

In [6]:
-- Check difference with offline
WITH AverageItemsOrdered AS (
    SELECT GENDER, AVG(NUMBER_ITEMS_OFFLINE) AS AvgItems
    FROM dbo.WOODCORP_ETC_CUSTOMERDATA AS C
    JOIN dbo.WOODCORP_ETC_OFFLINE AS O
    ON C.CUSTOMER_NO = O.CUSTOMER_NUMBER_OFFLINE
    GROUP BY GENDER
)
SELECT 
    (SELECT AvgItems FROM AverageItemsOrdered WHERE GENDER = 0) - 
    (SELECT AvgItems FROM AverageItemsOrdered WHERE GENDER = 1) AS DifferenceInAverageItems;

(1 row affected)

Total execution time: 00:00:00.088

DifferenceInAverageItems 3

Men tend to order 1 more item online and 3 more items offline compared to women. This could be indicative of wood production being a male-dominated industry, where men are more frequently responsible for placing larger orders. To address this gender-based difference, strategies should be developed to engage female customers more effectively. This could include targeted marketing campaigns that highlight the benefits and uses of wood products in areas that might appeal more to women, such as home improvement and crafts.

In [40]:
-- 3) What is the average offline order value for transporters which have more than 3 offline orders?
SELECT ROUND(AVG(ORDER_VALUE_OFFLINE), 0) AS AverageOfflineOrderValue
FROM dbo.WOODCORP_ETC_OFFLINE AS OFL
JOIN dbo.WOODCORP_ETC_CUSTOMERDATA AS C
ON OFL.CUSTOMER_NUMBER_OFFLINE = C.CUSTOMER_NO
WHERE C.CUSTOMER_TYPE = 1
AND C.CUSTOMER_NO IN (
    SELECT CUSTOMER_NUMBER_OFFLINE
    FROM dbo.WOODCORP_ETC_OFFLINE
    GROUP BY CUSTOMER_NUMBER_OFFLINE -- since we select customer mummer the counting of these means teh numer of orders 
    HAVING COUNT(*) > 3 -- The HAVING clause in SQL is used to filter records that are returned by a GROUP BY clause. It is similar to the WHERE clause, but HAVING is used for aggregate functions like COUNT, SUM, AVG, MIN, MAX, etc., while WHERE is used for filtering rows before the grouping occurs. here it ensures that only transporters who have placed more than three offline orders are considered in the calculation of the average offline order value.
);


(1 row affected)

Total execution time: 00:00:00.094

AverageOfflineOrderValue 4116

In [5]:
-- Check for online
SELECT ROUND(AVG(ORDER_VALUE_ONLINE), 0) AS AverageOnlineOrderValue
FROM dbo.WOODCORP_ETC_ONLINE AS OFL
JOIN dbo.WOODCORP_ETC_CUSTOMERDATA AS C
ON OFL.CUSTOMER_NUMBER_ONLINE = C.CUSTOMER_NO
WHERE C.CUSTOMER_TYPE = 1
AND C.CUSTOMER_NO IN (
    SELECT CUSTOMER_NUMBER_ONLINE
    FROM dbo.WOODCORP_ETC_ONLINE
    GROUP BY CUSTOMER_NUMBER_ONLINE -- since we select customer mummer the counting of these means teh numer of orders 
    HAVING COUNT(*) > 3
);

(1 row affected)

Total execution time: 00:00:00.054

AverageOnlineOrderValue 16596

Transporters with more than 3 orders spend significantly more on online orders (average value of 16,596) compared to offline orders (average value of 4,116). This suggests that transporters are more inclined to place larger orders online. To capitalize on this trend, businesses should focus on strengthening their online sales platform, providing targeted promotions and discounts for online bulk purchases, and improving the overall online purchasing experience for transporters.

In [41]:
-- 4) How many transporters order on average more than 2 discounted items online?
SELECT COUNT(*) AS TransportersWithMoreThan2DiscountedItems
FROM (
    SELECT CUSTOMER_NO, AVG(NUMBER_DISCOUNTED_ITEMS_ONLINE) AS AvgDiscountedItems
    FROM dbo.WOODCORP_ETC_ONLINE AS O
    JOIN dbo.WOODCORP_ETC_CUSTOMERDATA AS C
    ON O.CUSTOMER_NUMBER_ONLINE = C.CUSTOMER_NO
    WHERE C.CUSTOMER_TYPE = 1
    GROUP BY CUSTOMER_NO
    HAVING AVG(NUMBER_DISCOUNTED_ITEMS_ONLINE) > 2
) AS Transporters;


(1 row affected)

Total execution time: 00:00:00.055

TransportersWithMoreThan2DiscountedItems 763

In [14]:
-- CHeck for offline
SELECT COUNT(*) AS TransportersWithMoreThan2DiscountedItems
FROM (
    SELECT CUSTOMER_NO, AVG(NUMBER_DISCOUNT_ITEMS_OFFLINE) AS AvgDiscountedItems
    FROM dbo.WOODCORP_ETC_OFFLINE AS O
    JOIN dbo.WOODCORP_ETC_CUSTOMERDATA AS C
    ON O.CUSTOMER_NUMBER_OFFLINE = C.CUSTOMER_NO
    WHERE C.CUSTOMER_TYPE = 1
    GROUP BY CUSTOMER_NO
    HAVING AVG(NUMBER_DISCOUNT_ITEMS_OFFLINE) > 2
) AS Transporters;


(1 row affected)

Total execution time: 00:00:00.168

TransportersWithMoreThan2DiscountedItems 4049

Transporters show a higher tendency to order more than 2 discounted items offline (4,049) compared to online (763). This indicates that transporters are more responsive to discounts in offline transactions. To leverage this behavior, it is crucial to implement targeted discount campaigns for transporters in the offline market while exploring ways to replicate this success in online platforms, potentially through personalized discount offers and promotions.

In [43]:
-- 5) What age group has the highest online discount average for orders with more than 5 items?
SELECT TOP 1 AgeGroup, ROUND(AVG(DISCOUNT_VALUE_ONLINE), 0) AS AverageDiscount
FROM (
    SELECT CASE
        WHEN AGE = 1 THEN '0-18'
        WHEN AGE = 2 THEN '19-35'
        WHEN AGE = 3 THEN '36-55'
        WHEN AGE = 4 THEN '55+'
        WHEN AGE = 0 THEN 'NA' END AS AgeGroup,
        DISCOUNT_VALUE_ONLINE
    FROM dbo.WOODCORP_ETC_CUSTOMERDATA AS C
    JOIN dbo.WOODCORP_ETC_ONLINE AS O
    ON C.CUSTOMER_NO = O.CUSTOMER_NUMBER_ONLINE
    WHERE NUMBER_ITEMS_ONLINE > 5
) AS AgeDiscounts
GROUP BY AgeGroup
ORDER BY AverageDiscount DESC;


(1 row affected)

Total execution time: 00:00:00.061

AgeGroup AverageDiscount 36-55 2220

The age group 36-55 has the highest average online discount for orders with more than 5 items, at 2220. This suggests that this age group is particularly responsive to discounts, indicating an opportunity to tailor discount-focused marketing strategies towards them to maximize sales.

## **3) Additional Insights**

### **Additional Insight 1: Frequency of High-Quality Wood Orders Among Different Customer Types**

In [15]:
-- Determine which customer type orders high-quality wood most frequently offline
SELECT 
    CASE 
        WHEN C.CUSTOMER_TYPE = 1 THEN 'Transporter'
        WHEN C.CUSTOMER_TYPE = 2 THEN 'Factory'
        WHEN C.CUSTOMER_TYPE = 3 THEN 'Trader'
        WHEN C.CUSTOMER_TYPE = 4 THEN 'Private'
        ELSE 'Unknown'
    END AS CUSTOMER_TYPE,
    COUNT(CASE WHEN HIGH_QUALITY_WOOD > 0 THEN 1 END) AS HighQualityWoodOrders,
    COUNT(CASE WHEN LOW_QUALITY_WOOD > 0 THEN 1 END) AS LowQualityWoodOrders,
    COUNT(*) AS TotalWoodOrders,
    ROUND(COUNT(CASE WHEN HIGH_QUALITY_WOOD > 0 THEN 1 END) * 100.0 / COUNT(*), 2) AS HighQualityWoodProportion
FROM dbo.WOODCORP_ETC_OFFLINE AS O
JOIN dbo.WOODCORP_ETC_CUSTOMERDATA AS C
ON O.CUSTOMER_NUMBER_OFFLINE = C.CUSTOMER_NO
GROUP BY C.CUSTOMER_TYPE
ORDER BY HighQualityWoodOrders DESC;


Warning: Null value is eliminated by an aggregate or other SET operation.

(5 rows affected)

Total execution time: 00:00:00.090

CUSTOMER_TYPE HighQualityWoodOrders LowQualityWoodOrders TotalWoodOrders HighQualityWoodProportion Transporter 7798 8819 39941 19.520000000000 Factory 4580 4965 23539 19.460000000000 Trader 1485 1467 7411 20.040000000000 Private 1091 991 4873 22.390000000000 Unknown 414 453 1805 22.940000000000

Transporters have the highest frequency of high-quality wood orders (7,780), followed by factories (4,580). However, the proportion of high-quality wood orders relative to total orders is highest among private customers (22.34%), traders (20.06%), and factories (19.66%). This indicates that while transporters order the most high-quality wood in absolute terms, private customers and traders have a higher preference for high-quality wood relative to their total orders.

### **Additional Insight 2: Analysis of Discount Utilization by Customer Type**

In [16]:
-- Determine which customer type utilizes discounts the most in online orders
SELECT 
    CASE 
        WHEN CUSTOMER_TYPE = 1 THEN 'Transporter'
        WHEN CUSTOMER_TYPE = 2 THEN 'Factory'
        WHEN CUSTOMER_TYPE = 3 THEN 'Trader'
        WHEN CUSTOMER_TYPE = 4 THEN 'Private'
        ELSE 'Unknown'
    END AS CUSTOMER_TYPE,
    SUM(DISCOUNT_VALUE_ONLINE) AS TotalDiscountValue,
    SUM(ORDER_VALUE_ONLINE) AS TotalOrderValue,
    ROUND(SUM(DISCOUNT_VALUE_ONLINE) * 100.0 / SUM(ORDER_VALUE_ONLINE), 2) AS DiscountProportion
FROM dbo.WOODCORP_ETC_ONLINE AS O
JOIN dbo.WOODCORP_ETC_CUSTOMERDATA AS C
ON O.CUSTOMER_NUMBER_ONLINE = C.CUSTOMER_NO
GROUP BY CUSTOMER_TYPE
ORDER BY TotalDiscountValue DESC;


(5 rows affected)

Total execution time: 00:00:00.049

CUSTOMER_TYPE TotalDiscountValue TotalOrderValue DiscountProportion Transporter 6278844,410000007 48577951,86000014 12,93 Factory 5936052,78000002 49926713,45 11,89 Trader 1642637,8199999987 15086520,67999999 10,89 Private 559840,2899999997 5768434,3500000015 9,71 Unknown 158809,99000000002 1638136,6600000001 9,69

Transporters also lead in total discount value utilized in online orders (62,788,444), with factories and traders following. Transporters benefit the most from discounts (12.93% of their total order value), suggesting they are highly sensitive to price reductions.

Conclusion:

Given that transporters have the highest volume of high-quality wood orders and also benefit significantly from discounts, a combined strategy can be effective. Offering discounts on low-quality wood to transporters who frequently order high-quality wood can balance inventory and increase overall sales. Additionally, targeting private customers and traders with promotions for high-quality wood can leverage their higher relative interest in these products, while also offering them discounts on other products to encourage bulk purchasing. This dual approach can optimize customer satisfaction and sales performance across different customer segments.